dali/test/python/decoder/test_imgcodec.py

```diff
@@ -0,0 +1,407 @@
+    def pipe(device):
+        encoded, _ = fn.readers.file(file_root=file_root)
+        decoded = fn.experimental.decoders.image_crop(encoded, device=device, crop_w=10000, crop_h=100)
+        return decoded
+
+    p = pipe(device)
+    p.build()
+    assert_raises(RuntimeError, p.run,
+                  glob='cropping window*..*..*is not valid for image dimensions*[*x*]')
+
+
+def test_image_decoder_crop_error_oob():
+    for device in ['cpu', 'mixed']:
+        yield _testimpl_image_decoder_crop_error_oob, device
+
+
+def _testimpl_image_decoder_slice_error_oob(device):
+    file_root = os.path.join(test_data_root, good_path, "jpeg")
+
+    @pipeline_def(batch_size=batch_size_test, device_id=0, num_threads=4)
+    def pipe(device):
+        encoded, _ = fn.readers.file(file_root=file_root)
+        decoded = fn.experimental.decoders.image_slice(encoded, device=device, end=[10000], axes=[1])
+        return decoded
+
+    p = pipe(device)
+    p.build()
+    assert_raises(RuntimeError, p.run,
+                  glob='cropping window*..*..*is not valid for image dimensions*[*x*]')
+
+
+def test_image_decoder_slice_error_oob():
+    for device in ['cpu', 'mixed']:
+        yield _testimpl_image_decoder_slice_error_oob, device
+
+
+# TODO(msala) Implement HW decoder in nvJPEG
+
+# def test_pinned_input_hw_decoder():
+#     file_root = os.path.join(test_data_root, good_path, "jpeg")
+
+#     @pipeline_def(batch_size=128, device_id=0, num_threads=4)
+#     def pipe():
+#         encoded, _ = fn.readers.file(file_root=file_root)
+#         encoded_gpu = encoded.gpu()
+#         # encoded.gpu() should make encoded pinned
+#         decoded = fn.experimental.decoders.image(encoded, device="mixed")
+#         return decoded, encoded_gpu
+
+#     p = pipe()
+#     p.build()
+#     p.run()
+
+
+def test_tiff_palette():
+    normal = os.path.join(test_data_root, good_path, "tiff", "0/cat-300572_640.tiff")
+    palette = os.path.join(test_data_root, good_path, "tiff", "0/cat-300572_640_palette.tiff")
+
+    @pipeline_def(batch_size=2, device_id=0, num_threads=1)
+    def pipe():
+        encoded, _ = fn.readers.file(files=[normal, palette])
+        peeked_shapes = fn.peek_image_shape(encoded)
+        decoded = fn.experimental.decoders.image(encoded, device='cpu')
+        return decoded, peeked_shapes
+
+    p = pipe()
+    p.build()
+    imgs, peeked_shapes = p.run()
+    assert (peeked_shapes.at(0) == peeked_shapes.at(1)).all(), \
+        "Invalid peeked shape of palette TIFF"
+
+    delta = np.abs(imgs.at(0).astype('float') - imgs.at(1).astype('float'))/256
+    assert np.quantile(delta, 0.9) < 0.05, "Original and palette TIFF differ significantly"
+
+
+def _testimpl_image_decoder_peek_shape(path, expected_shape):
+    @pipeline_def(batch_size=1, device_id=0, num_threads=1)
+    def peek_shape_pipeline(file):
+        encoded, _ = fn.readers.file(files=[file])
+        return fn.peek_image_shape(encoded)
```

Comment:
Shouldn't that be fn.experimental.peek_image_shape?

Reply:
xD

Reply:
Thanks for spotting that! Now everything has exploded and I need some time to fix that :P

Reply:
Changed and fixed, thank you again!